src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,692 @@
+                    None
+                }
+                CustomTarget::FinalBin => {
+                    // Visit all dependencies but do not add this unit to
+                    // result.
+                    for dep in &units[i].dependencies {
+                        visit(units, result, visited, dep.index);
+                    }
+                    None
+                }
+            };
+            visited[i] = Some(r);
+            r
+        }
+
+        for root in &self.roots {
+            visit(&self.units, &mut result, &mut visited, *root);
+        }
+
+        result
+    }
+}
+
+/// A representation of a command that can be easily formatted as a shell
+/// command or turned into an actual [Command].
+struct CmdInvocation {
+    program: String,
+    args: Vec<String>,
+    env: Vec<(String, String)>,
+}
+
+impl CmdInvocation {
+    fn as_command(&self) -> Command {
+        let mut cmd = Command::new(&self.program);
+        cmd.args(&self.args)
+            .envs(self.env.iter().map(|(k, v)| (k, v)));
+        cmd
+    }
+}
+
+impl fmt::Display for CmdInvocation {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for (k, v) in &self.env {
+            write!(f, "{}={} ", k, escape(v.into()))?;
+        }
+        write!(f, "{}", self.program)?;
+        for arg in &self.args {
+            write!(f, " {}", escape(arg.into()))?;
+        }
+        Ok(())
+    }
+}
+
+fn main() {
+    let target_triple = env::var("TARGET")
+        .unwrap_or_else(|_| rustc_session::config::host_triple().into());
+    let cwd: Utf8PathBuf = env::current_dir()
+        .expect("cwd should be accessible")
+        .try_into()
+        .expect("cwd should be UTF-8");
+    let custom_sources_dir = match env::var_os("MIR_JSON_RUST_SOURCES_PATH") {
+        Some(path) => fs::canonicalize(path)
+            .expect("MIR_JSON_RUST_SOURCES_PATH should be a valid path")
+            .try_into()
+            .expect("MIR_JSON_RUST_SOURCES_PATH should be UTF-8"),
+        None => cwd.join("libs"),
+    };
+    let generate_only = env::args().skip(1).any(|arg| arg == "--generate");
+
+    // Compute the paths that we will use later to store the build artifacts,
+    // but don't actually create them yet.
+    eprintln!("Querying paths...");
+    let rlibs_sysroot = cwd.join("rlibs_real");
+    let rlibs_symlink = cwd.join("rlibs");
+    let rlibs_dir = Command::new("rustc")
+        .arg("--target")
+        .arg(&target_triple)
+        .arg("--print")
+        .arg("target-libdir")
+        .arg("--sysroot")
+        .arg(&rlibs_sysroot)
+        .output()
+        .expect("rustc --print target-libdir should run")
+        .stdout;
+    let rlibs_dir: &Utf8Path = std::str::from_utf8(&rlibs_dir)
+        .expect("rustc target-libdir should be UTF-8")
+        .trim_end()
+        .into();
+
+    // Set up a new cargo project for running `cargo test -Z build-std`.
+    eprintln!("Creating empty cargo package...");
```

Comment:
It might be worth creating this project in a temporary directory using [`tempdir`](https://docs.rs/tempfile/3.19.1/tempfile/fn.tempdir.html). This would avoid needing to have to delete it beforehand in the event of an error, and it would automatically handle cleaning up the project when the `TempDir` is dropped.

Reply:
Good idea!